### PR TITLE
[copy] Fix copying file with name ending in slash

### PR DESCRIPTION
**Pull request by jigold**
- Also handle case with GCS files ending in a slash

### CODE REVIEW

File: hail/python/hailtop/aiogoogle/client/storage_client.py [hunk 1]
```diff
@@ -652,7 +652,14 @@ async def _listfiles_recursive(self, bucket: str, name: str) -> AsyncIterator[Fi
             items = page.get('items')
             if items is not None:
                 for item in page['items']:
-                    yield GoogleStorageFileListEntry(f'gs://{bucket}/{item["name"]}', item)
+                    url = f'gs://{bucket}/{item["name"]}'
+                    if url.endswith('/') and item is not None:
```
Comment (cseed) [inline hail/python/hailtop/aiogoogle/client/storage_client.py:656]: `item` cannot be None: we subscript it in the line above.

I'd write this like this:

```
  entry = GoogleStorageFileListEntry(url, item)
  if url.endswith('/'):
    status = await entry.status()
    size = await entry.size()
    if size != 0:
      raise FileAndDirectoryError(url)
  yield entry
```

The entry caches the status, so this way, if the user gets the status, it reuses the status object you create.

File: hail/python/hailtop/aiotools/fs.py [hunk 2]
```diff
@@ -634,6 +639,10 @@ async def copy_as_dir(self, sema: asyncio.Semaphore, source_report: SourceReport
             self.src_is_dir = False
             await self.release_barrier()
             return
+        except:
+            self.encountered_error = True
+            await self.release_barrier()
+            raise
```
Comment (cseed) [inline hail/python/hailtop/aiotools/fs.py:645]: I think I'd write this like this:

```
    async def copy_as_dir(self, sema: asyncio.Semaphore, source_report: SourceReport, return_exceptions: bool):
        try:
	    src = self.src
	    if not src.endswith('/'):
		src = src + '/'
	    try:
		srcentries = await self.router_fs.listfiles(src, recursive=True)
	    except (NotADirectoryError, FileNotFoundError):
		self.src_is_dir = False
		return
	    self.src_is_dir = True
        finally:
            await self.release_barrier()

        await self.barrier.wait()

        if self.src_is_file is not None and self.src_is_file:
            raise FileAndDirectoryError(self.src)

        ...
```

This makes the release of the barrier clearer.  Then we can remove release_barrier_and_wait.  We need the `self.src_is_file is not None` in the case `copy_as_file` got an error.

File: hail/python/hailtop/aiotools/fs.py [hunk 3]
```diff
@@ -673,11 +682,13 @@ async def copy(self, sema: asyncio.Semaphore, source_report: SourceReport, retur
                 return_exceptions=True)
 
             assert self.pending == 0
-            assert (self.src_is_file is None) == self.src.endswith('/')
-            assert self.src_is_dir is not None
 
-            if (self.src_is_file is False or self.src.endswith('/')) and not self.src_is_dir:
-                raise FileNotFoundError(self.src)
+            if not self.encountered_error:
```
Comment (cseed) [inline hail/python/hailtop/aiotools/fs.py:686]: Then you can put the check for errors above these checks, and get rid of encountered_error.

File: hail/python/test/hailtop/aiotools/test_copy.py [hunk 4]
```diff
@@ -416,3 +417,22 @@ async def test_copy_src_parts(copy_test_context):
...
+async def test_file_and_directory_error_with_slash(router_filesystem):
+    sema, fs, bases = router_filesystem
+
+    src_base = await fresh_dir(fs, bases, 'gs')
+    dest_base = await fresh_dir(fs, bases, 'gs')
+
+    await fs.create(f'{src_base}empty/')
+
+    async with await fs.create(f'{src_base}not-empty/') as f:
+        await f.write('foo'.encode('utf-8'))
```
Comment (cseed) [inline hail/python/test/hailtop/aiotools/test_copy.py:432]: Or just `b'foo'`.

File: hail/python/test/hailtop/aiotools/test_copy.py [hunk 5]
```diff
@@ -416,3 +417,22 @@ async def test_copy_src_parts(copy_test_context):
...
+
+    with pytest.raises(FileNotFoundError):
+        await fs.copy(sema, Transfer(f'{src_base}empty/', dest_base.rstrip('/')))
```
Comment (cseed) [inline hail/python/test/hailtop/aiotools/test_copy.py:435]: Might also be nice to check this works as expected when the empty file isn't at the toplevel.  Like, have src_base/a/{foo, empty/} and copy src_base/a and make sure foo copies but not empty/.

File: hail/python/test/hailtop/aiotools/test_copy.py [hunk 6]
```diff
@@ -416,3 +417,22 @@ async def test_copy_src_parts(copy_test_context):
...
+
+    with pytest.raises(FileAndDirectoryError):
+        await fs.copy(sema, Transfer(f'{src_base}not-empty/', dest_base.rstrip('/')))
```
Comment (cseed) [inline hail/python/test/hailtop/aiotools/test_copy.py:438]: Same, not at top level.

File: hail/python/hailtop/aiotools/fs.py [hunk 7]
```diff
@@ -608,6 +609,10 @@ async def copy_as_file(self,
             self.src_is_file = False
             await self.release_barrier()
             return
+        except:
```
Comment (cseed) [inline hail/python/hailtop/aiotools/fs.py:612]: Can I ask, what unhandled error did we see here?  Just so I can keep this case in mind in the future.

**Review by cseed — CHANGES_REQUESTED** [review]

File: hail/python/hailtop/aiotools/fs.py (hunk 7 above)
Comment (jigold) [reply hail/python/hailtop/aiotools/fs.py:612]: Lindo's service account didn't have permissions to access the files. So he got a google 403 exception. But the error message was `assert pending == 0`

File: hail/python/test/hailtop/aiotools/test_copy.py [hunk 8]
```diff
@@ -416,3 +417,34 @@ async def test_copy_src_parts(copy_test_context):
...
+    await fs.create(f'{src_base}a/empty/')
+    await fs.create(f'{src_base}a/foo')
+
+    async with await fs.create(f'{src_base}b/not-empty/') as f:
+        await f.write(b'not-empty')
+
+    async with await fs.create(f'{src_base}b/bar') as f:
+        await f.write(b'bar')
+
+    await fs.copy(sema, Transfer(f'{src_base}a/foo', dest_base.rstrip('/')))
```
Comment (cseed) [inline hail/python/test/hailtop/aiotools/test_copy.py:438]: What's this testing?

File: hail/python/test/hailtop/aiotools/test_copy.py [hunk 9]
```diff
@@ -416,3 +417,34 @@ async def test_copy_src_parts(copy_test_context):
...
+
+@pytest.mark.asyncio
+async def test_file_and_directory_error_with_slash(router_filesystem):
+    sema, fs, bases = router_filesystem
+
+    src_base = await fresh_dir(fs, bases, 'gs')
+    dest_base = await fresh_dir(fs, bases, 'gs')
+
+    await fs.create(f'{src_base}a/empty/')
+    await fs.create(f'{src_base}a/foo')
```
Comment (cseed) [inline hail/python/test/hailtop/aiotools/test_copy.py:430]: Sorry, my previous comments weren't clear.  Let me try again:

I'm imagining you have:

 - {src_base}/empty/ (which is an empty file) and {src_base}/empty/foo
 - {src_base}/non-empty/ (which is a non-empty file) and {src_base}/non-empty/bar
 - {src_base}/empty-only/  (which is empty, but empty-only has no other files in it)

Then copy empty, and verify that the empty directory file _wasn't_ copied, and foo was.

Then verify copying from `non-empty/` gets a FileAndDirectoryError.

Then copy empty-only/.  I think this should work (it is a directory).  If you're copying to the local filesystem, it should create the target directory (I suspect we're not doing that now).

```
$ mkdir scratch; cd scratch
$ mkdir a
$ cp -a a b
$ find .
.
./b
./a
```

**Review by cseed — CHANGES_REQUESTED** [review]
Made comment.

File: hail/python/test/hailtop/aiotools/test_copy.py [hunk 10]
```diff
@@ -416,3 +417,40 @@ async def test_copy_src_parts(copy_test_context):
...
+    await fs.copy(sema, Transfer(f'{src_base}not-empty/bar', dest_base.rstrip('/')))
+    await expect_file(fs, f'{dest_base}not-empty/bar', 'bar')
+
+    with pytest.raises(FileAndDirectoryError):
+        await fs.copy(sema, Transfer(f'{src_base}not-empty/', dest_base.rstrip('/')))
+
+    # Test empty file that ends in a slash
+    await fs.create(f'{src_base}empty-only/')
+
+    with pytest.raises(FileNotFoundError):
```
Comment (cseed) [inline hail/python/test/hailtop/aiotools/test_copy.py:455]: I think this shouldn't be FileNotFound.  {src_base}empty_only/ exists and is a directory.  This should work, see my cp example in another comment.

File: hail/python/test/hailtop/aiotools/test_copy.py [hunk 11]
```diff
@@ -416,3 +417,40 @@ async def test_copy_src_parts(copy_test_context):
...
+
+
+@pytest.mark.asyncio
+async def test_file_and_directory_error_with_slash(router_filesystem):
+    sema, fs, bases = router_filesystem
+
+    src_base = await fresh_dir(fs, bases, 'gs')
+    dest_base = await fresh_dir(fs, bases, 'gs')
+
+    # test empty file ending in slash and directory
```
Comment (cseed) [inline hail/python/test/hailtop/aiotools/test_copy.py:429]: OCD nit: this has a lower case t [t]est, while the comments below have upper case Ts.

**Review by cseed — CHANGES_REQUESTED** [review]

File: hail/python/test/hailtop/aiotools/test_copy.py (hunk 9 above)
Comment (jigold) [reply hail/python/test/hailtop/aiotools/test_copy.py]: This is not working. The case when copying `{src_base}/empty/foo` results in FileNotFound as a Google Exception.

Comment (jigold) [reply hail/python/test/hailtop/aiotools/test_copy.py]: Wait, sorry the Google Exception happens in the test's assert, not in the copy code.

**Comment by jigold** [comment]
I'm going to split this into two PRs and I'll figure out why the file ending in a slash code isn't working when I get back.

File: hail/python/test/hailtop/aiotools/test_copy.py [hunk 12]
```diff
@@ -416,3 +417,38 @@ async def test_copy_src_parts(copy_test_context):
...
+
+    await fs.copy(sema, Transfer(f'{src_base}not-empty/bar', dest_base.rstrip('/')))
+    await expect_file(fs, f'{dest_base}not-empty/bar', 'bar')
+
+    with pytest.raises(FileAndDirectoryError):
+        await fs.copy(sema, Transfer(f'{src_base}not-empty/', dest_base.rstrip('/')))
+
+    # Test empty file that ends in a slash
+    await fs.create(f'{src_base}empty-only/')
+    await fs.copy(sema, Transfer(f'{src_base}empty-only/', dest_base.rstrip('/')))
```
Comment (cseed) [inline hail/python/test/hailtop/aiotools/test_copy.py:454]: You should do this with a dest that doesn't exist, and verify it gets created.

**Review by cseed — CHANGES_REQUESTED** [review]

File: hail/python/test/hailtop/aiotools/test_copy.py [hunk 13]
```diff
@@ -416,3 +417,38 @@ async def test_copy_src_parts(copy_test_context):
...
+
+    # Test empty file ending in slash and directory
+    await fs.create(f'{src_base}empty/')
+    async with await fs.create(f'{src_base}empty/foo') as f:
+        await f.write(b'foo')
+
+    await fs.copy(sema, Transfer(f'{src_base}empty/', dest_base.rstrip('/')))
+    await expect_file(fs, f'{dest_base}empty/foo', 'foo')
+    with pytest.raises(AssertionError):
+        await expect_file(fs, f'{dest_base}empty/', '')
```
Comment (cseed) [inline hail/python/test/hailtop/aiotools/test_copy.py:437]: Don't expect_file something that doesn't exist.  Just do `assert not fs.exists(file)`.

File: hail/python/test/hailtop/aiotools/test_copy.py (hunk 12 above)
Comment (jigold) [reply hail/python/test/hailtop/aiotools/test_copy.py:453]: It already doesn't exist. It's `empty-only` which is different than `empty`

**Comment by jigold** [comment]
I don't think these tests will pass. I'll address the issues when I come back from vacation. Feel free to reopen if you want.

File: hail/python/test/hailtop/aiotools/test_copy.py (hunk 12 above)
Comment (cseed) [reply hail/python/test/hailtop/aiotools/test_copy.py:453]: The code creates dest_base, see copy_test_context:

```
    # make sure dest_base exists
    async with await fs.create(f'{dest_base}keep'):
        pass
```

**Comment by jigold** [comment]
Do I need to do something on my end in order to test copying from my computer rather than through CI? I keep getting a 400 regardless of whether I use my bucket `hail-jigold` or the test bucket `hail-test-dmk9z`. The only information I can get is "Bad Request" which I think is from this:

> The request cannot be completed based on your current Cloud Storage settings. For example, you cannot lock a retention policy if the requested bucket doesn't have a retention policy, and you cannot set ACLs if the requested bucket has Bucket Policy Only enabled.

I don't think this is the real error as we know copying works fine with the Hail test bucket.

File: hail/python/hailtop/aiogoogle/client/storage_client.py [hunk 14]
```diff
@@ -423,11 +424,10 @@ async def __getitem__(self, key: str) -> str:
 
 
 class GoogleStorageFileListEntry(FileListEntry):
-    def __init__(self, url: str, items: Optional[Dict[str, Any]]):
-        assert url.endswith('/') == (items is None), f'{url} {items}'
+    def __init__(self, url: str, items: Optional[Dict[str, Any]], status=None):
```
Comment (cseed) [inline hail/python/hailtop/aiogoogle/client/storage_client.py:427]: Type hint on `status`

File: hail/python/hailtop/aiogoogle/client/storage_client.py [hunk 15]
```diff
@@ -644,9 +644,17 @@ async def _listfiles_recursive(self, bucket: str, name: str) -> AsyncIterator[Fi
             items = page.get('items')
             if items is not None:
                 for item in page['items']:
-                    yield GoogleStorageFileListEntry(f'gs://{bucket}/{item["name"]}', item)
-
-    async def _listfiles_flat(self, bucket: str, name: str) -> AsyncIterator[FileListEntry]:
+                    url = f'gs://{bucket}/{item["name"]}'
+                    status = None
+                    if not force_all and url.endswith('/') and item is not None:
```
Comment (cseed) [inline hail/python/hailtop/aiogoogle/client/storage_client.py:649]: all is a bit general when you're just controlling empty directory files.  What about `skip_empty_directory_files` (or `include_empty_directory_files` if you want the positive sense, although personally I find the sense that doesn't require a `not` to be more readable.)

File: hail/python/hailtop/aiogoogle/client/storage_client.py [hunk 16]
```diff
@@ -659,22 +667,34 @@ async def _listfiles_flat(self, bucket: str, name: str) -> AsyncIterator[FileLis
                 for prefix in prefixes:
                     assert prefix.endswith('/')
                     url = f'gs://{bucket}/{prefix}'
+                    if not force_all and await self.isfile(url):
```
Comment (cseed) [inline hail/python/hailtop/aiogoogle/client/storage_client.py:670]: I think this is wrong.  We should return directories that contain only empty files.

File: hail/python/hailtop/aiogoogle/client/storage_client.py (hunk 15 above)
Comment (cseed) [reply hail/python/hailtop/aiogoogle/client/storage_client.py:649]: Also, I don't think `item is not None` is needed here.  Item shouldn't be able to be None (let's add an assertion if we're unsure).

File: hail/python/hailtop/aiotools/fs.py [hunk 17]
```diff
@@ -632,11 +632,18 @@ async def copy_as_dir(self, sema: asyncio.Semaphore, source_report: SourceReport
...
+            srcentries = None
+
             try:
                 srcentries = await self.router_fs.listfiles(src, recursive=True)
-            except (NotADirectoryError, FileNotFoundError):
-                self.src_is_dir = False
-                return
+            except (NotADirectoryError, FileNotFoundError) as e:
+                if isinstance(e, FileNotFoundError):
+                    if await self.router_fs.isdir(src):
```
Comment (cseed) [inline hail/python/hailtop/aiotools/fs.py:642]: If possible, I really don't want to have to do additional API requests after getting the listings.  So we should make the listfiles interface operate in a way that lets us do the right thing here without more calls.

In addition to your questions, another one is how listfiles behaves in the presence of empty directory files.  I think listfiles shouldn't throw NotADirectoryError/FileNotFoundError if the directory only has an empty directory file.

**Review by cseed — CHANGES_REQUESTED** [review]

File: hail/python/hailtop/aiogoogle/client/storage_client.py (hunk 16 above)
Comment (jigold) [reply hail/python/hailtop/aiogoogle/client/storage_client.py:670]: See my comment above. For `rmtree` where `force_all` is True, then we do want to return non-empty directory/file files.

File: hail/python/hailtop/aiotools/fs.py [hunk 18]
```diff
@@ -665,9 +672,12 @@ async def copy_source(srcentry):
 
             await self._copy_file_multi_part(sema, source_report, srcfile, await srcentry.status(), url_join(full_dest, relsrcfile), return_exceptions)
 
-        await bounded_gather2(sema, *[
-            copy_source(srcentry)
-            async for srcentry in srcentries], cancel_on_error=True)
+        if srcentries is None:
```
Comment (jigold) [inline hail/python/hailtop/aiotools/fs.py:675]: I'm uneasy with the empty directory creation. Even if we handle this use case where a top level directory is an empty file/directory, there's still downstream empty file/directory being missed completely and we won't make those empty directories on the local file system where mkdir is not a noop.

For example,

path/foo/ is the only match

versus.

path/bar/foo/ and the query is path/bar/.

File: hail/python/hailtop/aiogoogle/client/storage_client.py (hunk 15 above)
Comment (cseed) [reply hail/python/hailtop/aiogoogle/client/storage_client.py:649]: The reason I don't like `all` is that it isn't clear what files are excluded when you set `include_all_files=False`.

**Review by cseed — CHANGES_REQUESTED** [review]
Getting close, some additional suggestions to simplify the code.

File: hail/python/test/hailtop/aiotools/test_copy.py [hunk 19]
```diff
@@ -416,3 +417,135 @@ async def test_copy_src_parts(copy_test_context):
...
+    await fs.listfiles(f'{src_base}empty-only/')
+    await fs.listfiles(f'{src_base}empty-only/', recursive=True)
+
+    for transfer_type in (Transfer.DEST_IS_TARGET, Transfer.DEST_DIR, Transfer.INFER_DEST):
+        dest_base = await fresh_dir(fs, bases, 'gs')
+        await fs.copy(sema, Transfer(f'{src_base}empty-only/', dest_base.rstrip('/'), treat_dest_as=transfer_type))
+
+        # We ignore empty directories when copying
+        with pytest.raises(FileNotFoundError):
+            await fs.listfiles(f'{dest_base}empty-only/')
```
Comment (cseed) [inline hail/python/test/hailtop/aiotools/test_copy.py:516]: I think this one is wrong now?

File: hail/python/test/hailtop/aiotools/test_copy.py [hunk 20]
```diff
@@ -416,3 +417,135 @@ async def test_copy_src_parts(copy_test_context):
...
+
+    src_base = await fresh_dir(fs, bases, 'gs')
+
+    await write_file(fs, f'{src_base}empty/', '')
+    await write_file(fs, f'{src_base}empty/foo', b'foo')
+
+    await fs.listfiles(f'{src_base}')
+    await fs.listfiles(f'{src_base}', recursive=True)
+    await fs.listfiles(f'{src_base}empty/')
+    await fs.listfiles(f'{src_base}empty/', recursive=True)
```
Comment (cseed) [inline hail/python/test/hailtop/aiotools/test_copy.py:439]: `listfiles` returns an async generator, and you won't test the full behavior unless you drain the generator.

**Review by cseed — DISMISSED** [review]
Code is looking good!  I know the back and forth is tedious, but I think the chance turned out nice in the end.

File: hail/python/hailtop/aiogoogle/client/storage_client.py [hunk 21]
```diff
@@ -682,11 +681,27 @@ async def listfiles(self, url: str, recursive: bool = False) -> AsyncIterator[Fi
         except StopAsyncIteration:
             raise FileNotFoundError(url)  # pylint: disable=raise-missing-from
 
+        async def is_file(entry):
+            if await entry.is_dir():
+                return False
```
Comment (cseed) [inline hail/python/hailtop/aiogoogle/client/storage_client.py:686]: So I think `is_file` is a bad name.  This should be something like `skip` or `should_yield`?  Because you want to return directories here.  You only want to skip empty-file-directories.

Comment (jigold) [reply hail/python/hailtop/aiogoogle/client/storage_client.py:686]: Why would I want to return directories? We're listing files (or at least that's what I thought it was). 

I also broke listfiles in other tests:

```
        async def listfiles(dir, recursive):
            return {(await entry.url(), await entry.is_file()) async for entry in await fs.listfiles(dir, recursive)}
    
        assert await listfiles(f'{base}foo/', recursive=True) == {(a, True), (c, True)}
>       assert await listfiles(f'{base}foo/', recursive=False) == {(a, True), (b, False)}
E       AssertionError: assert {('gs://hail-test-dmk9z/tmp/QPTOQ0djdUWM3WYNHTbKew/foo/a', True)} == {('gs://hail-test-dmk9z/tmp/QPTOQ0djdUWM3WYNHTbKew/foo/a', True),\n ('gs://hail-test-dmk9z/tmp/QPTOQ0djdUWM3WYNHTbKew/foo/b/', False)}
E         Extra items in the right set:
E         ('gs://hail-test-dmk9z/tmp/QPTOQ0djdUWM3WYNHTbKew/foo/b/', False)
E         Full diff:
E           {
E            ('gs://hail-test-dmk9z/tmp/QPTOQ0djdUWM3WYNHTbKew/foo/a',
E             True),
E         -  ('gs://hail-test-dmk9z/tmp/QPTOQ0djdUWM3WYNHTbKew/foo/b/',
E         -   False),
E           }
```

```
        assert await listfiles(f'{base}foo/', recursive=True) == {(a, True), (c, True)}
>       assert await listfiles(f'{base}foo/', recursive=False) == {(a, True), (b, False)}
E       AssertionError: assert {('gs://hail-test-dmk9z/tmp/fOxVwBRMRHj37hvn8H20i1/foo/a', True)} == {('gs://hail-test-dmk9z/tmp/fOxVwBRMRHj37hvn8H20i1/foo/a', True),\n ('gs://hail-test-dmk9z/tmp/fOxVwBRMRHj37hvn8H20i1/foo/b/', False)}
E         Extra items in the right set:
E         ('gs://hail-test-dmk9z/tmp/fOxVwBRMRHj37hvn8H20i1/foo/b/', False)
E         Full diff:
E           {
E            ('gs://hail-test-dmk9z/tmp/fOxVwBRMRHj37hvn8H20i1/foo/a',
E             True),
E         -  ('gs://hail-test-dmk9z/tmp/fOxVwBRMRHj37hvn8H20i1/foo/b/',
E         -   False),
E           }

```

Comment (jigold) [reply hail/python/hailtop/aiogoogle/client/storage_client.py:686]: Maybe these tests need to be fixed rather than my code being wrong.

Comment (cseed) [reply hail/python/hailtop/aiogoogle/client/storage_client.py:686]: Hmm, no, the code is definitely wrong.  `listfiles(..., recursive=False)` should definitely be returning **directories**.

Comment (cseed) [reply hail/python/hailtop/aiogoogle/client/storage_client.py:686]: Let me clarify a bit:

`listfiles(..., recursive=True)` automatically descends into directories and only returns files.

`listfiles(..., recursive=False)` acts like listing directories on UNIX, and only returns things directly inside a directory, whether it be files or more directories.

Google Storage listing API supports both of these modes directly.  UNIX directory listing only implements `recursive=False`, and we iterate recursively through directories explicit in the LocalAsyncFS backend.

We don't make much use of `recursive=False`, but I think users will want it when we expose this to them (replacing hl.hadoop_ls, etc.)

Does that help?

Comment (cseed) [reply hail/python/hailtop/aiogoogle/client/storage_client.py:686]: Obviously, I need to document the AsyncFS API ASAP.

Comment (cseed) [reply hail/python/hailtop/aiogoogle/client/storage_client.py:686]: And maybe rename `listfiles` to `listdir`?  We're listing the contents of a directory.

Comment (jigold) [reply hail/python/hailtop/aiogoogle/client/storage_client.py]: I think the equivalent operations of what we're doing is listfiles_flat is the equivalent of scandir in Python. listfiles_recursive doesn't have an equivalent operation in unix or python. I looked at `ls -R` but that lists both file and directory entries. `listfiles` makes sense to me for `listfiles_recursive`. `listdir` makes sense to me for `listfiles_flat`. Can we have both operations be top-level things rather than the `recursive=True` flag?

Comment (cseed) [reply hail/python/hailtop/aiogoogle/client/storage_client.py]: > Can we have both operations be top-level things rather than the recursive=True flag?

Yeah, definitely.  I'll put it on the list of things to cleanup and refine when I start documenting.

**Comment by jigold** [comment]
bump

File: hail/python/hailtop/aiogoogle/client/storage_client.py [hunk 22]
```diff
@@ -424,7 +425,6 @@ async def __getitem__(self, key: str) -> str:
 
 class GoogleStorageFileListEntry(FileListEntry):
     def __init__(self, url: str, items: Optional[Dict[str, Any]]):
-        assert url.endswith('/') == (items is None), f'{url} {items}'
```
[the remark this answers is not in the document]
Comment (cseed) [reply hail/python/hailtop/aiogoogle/client/storage_client.py:427]: Why did you remove this?  This assertion is still good, right?

**Review by cseed — DISMISSED** [review]

File: hail/python/hailtop/aiogoogle/client/storage_client.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (jigold) [reply hail/python/hailtop/aiogoogle/client/storage_client.py:427]: No. It's basically representing a directory if items is None. We now have directory entries in our list, but we filter them out in `listfiles`. This is the central issue that needed to be fixed (user would hit this assertion error when they had both a file and a directory).

https://github.com/hail-is/hail/blob/f408e2734483befad8aaba9bb4ef2df5f017a161/hail/python/hailtop/aiogoogle/client/storage_client.py#L662

[the remark this answers is not in the document]
Comment (cseed) [reply hail/python/hailtop/aiogoogle/client/storage_client.py:427]: Ah, got it, thanks for clarifying!

**Review by cseed — APPROVED** [review]
